### PR TITLE
initial yaml parse for cli

### PR DESCRIPTION
**Pull request by d-g-town**

### CODE REVIEW

**Review by ianedwards — APPROVED** [review]
left some comments but lgtm! there are some prints to watch out for

File: api/server/handlers/porter_app_v2/parse_yaml.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	fmt.Println("1")
+
+	if !project.ValidateApplyV2 {
+		err := telemetry.Error(ctx, span, nil, "project does not have apply v2 enabled")
+		c.HandleAPIError(w, r, apierrors.NewErrForbidden(err))
+		return
+	}
+
+	fmt.Println("2")
+
```
[the remark this answers is not in the document]
Comment (d-g-town) [reply api/server/handlers/porter_app_v2/parse_yaml.go:62]: thank you!

File: api/server/router/porter_app_v2.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	relPath := "/appv2"
+
+	newPath := &types.Path{
+		Parent:       basePath,
+		RelativePath: relPath,
+	}
+
+	var routes []*router.Route
+
+	// GET /api/projects/{project_id}/clusters/{cluster_id}/appv2/parse -> porter_app_v2_v2.NewParsePorterYAMLToProtoHandler
```
[the remark this answers is not in the document]
Comment (d-g-town) [reply api/server/router/porter_app_v2.go:56]: it was a bad motivation, will slot it into `/app` path inside the `porter_app` router

**Review by jose-fully-ported — APPROVED** [review]

File: api/server/handlers/porter_app/parse_yaml.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+	request := &ParsePorterYAMLToProtoRequest{}
+	if ok := c.DecodeAndValidate(w, r, request); !ok {
+		err := telemetry.Error(ctx, span, nil, "error decoding request")
+		c.HandleAPIError(w, r, apierrors.NewErrPassThroughToClient(err, http.StatusBadRequest))
+		return
+	}
+
+	if request.Base64Yaml == "" {
+		err := telemetry.Error(ctx, span, nil, "b64 yaml is empty")
```
Comment (stefanmcshane) [inline api/server/handlers/porter_app/parse_yaml.go:68]: b64 or base64 for consistency

File: cli/cmd/apply.go [hunk 4]
```diff
@@ -120,8 +120,9 @@ func apply(_ *types.GetAuthenticatedUserResponse, client *api.Client, _ []string
 	}
 
 	if project.ValidateApplyV2 {
-		err = v2.Apply(ctx)
+		err = v2.Apply(ctx, cliConf, client, porterYAML)
 		if err != nil {
+			color.New(color.FgRed).Fprintf(os.Stderr, "Error: %s\n", err.Error()) // nolint:errcheck,gosec
```
Comment (stefanmcshane) [inline cli/cmd/apply.go:125]: What was the gosec error?

File: internal/porter_app/parse.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	version := &PorterYAMLVersion{}
+	err := yaml.Unmarshal(porterYaml, version)
+	if err != nil {
+		return nil, telemetry.Error(ctx, span, err, "error unmarshaling porter yaml")
+	}
+
+	var appProto *porterv1.PorterApp
+	switch version.Version {
```
Comment (stefanmcshane) [inline internal/porter_app/parse.go:35]: There are a few places around the codebase that we check a file called porter.yaml's version. Can we make sure that they will handle the v2 case also where necessary

File: internal/porter_app/parse.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	case "v2":
```
Comment (stefanmcshane) [inline internal/porter_app/parse.go:36]: Consider making these a constant which can be imported around the API 

```go
type ApplicationVersion string
const (
    ApplicationVersion_V2 ApplicationVersion = v1
)
```

**Review by stefanmcshane — APPROVED** [review]

File: cli/cmd/apply.go (hunk 4 above)
Comment (d-g-town) [reply cli/cmd/apply.go:125]:
```
cli/cmd/apply.go:124:4: G104: Errors unhandled. (gosec)
                        color.New(color.FgRed).Fprintf(os.Stderr, "Error: %s\n", err.Error()) // nolint:errcheck
```
Seemed strange to me that it wouldn't let errcheck handle this.

File: internal/porter_app/parse.go (hunk 5 above)
Comment (d-g-town) [reply internal/porter_app/parse.go:35]: Yes, the other places I'm aware of that deal with the version are the old apply flow (will error if the version is not "v1", "v2beta", "v1stack" or "") and getPorterYaml endpoint (will error if version is not "v1stack")

**Pull request merged**